tests/garage/tf/optimizers/test_conjugate_gradient_optimizer.py

```diff
@@ -0,0 +1,240 @@
+"""Tests for garage.tf.optimizers.conjugateGradientOptimizer"""
+import numpy as np
+import pytest
+import tensorflow as tf
+
+from garage.tf.optimizers.conjugate_gradient_optimizer import (
+    cg, FiniteDifferenceHvp, PerlmutterHvp)
+from garage.tf.policies import Policy
+from tests.fixtures import TfGraphTestCase
+
+
+class HelperPolicy(Policy):
+    """Helper policy class for testing hvp classes"""
+
+    def __init__(self, n_vars, name='OneParamPolicy'):
+        super().__init__(name, None)
+        with tf.compat.v1.variable_scope(self.name) as vs:
+            self._variable_scope = vs
+            _ = [tf.Variable([0.]) for _ in range(n_vars)]
+
+    def get_action(self, observation):
+        pass
+
+    def get_actions(self, observations):
+        pass
+
+
+class TestConjugateGradientOptimizer:
+    """Test class for ConjugateGradientOptimizer and HVP classes"""
+
+    def test_cg(self):
+        """Solve Ax = b using Conjugate gradient method."""
+        a = np.linspace(-np.pi, np.pi, 25).reshape((5, 5))
+        a = a.T.dot(a)  # make sure a is positive semi-definite
+        b = np.linspace(-np.pi, np.pi, 5)
+        x = cg(a.dot, b, cg_iters=5)
+        assert np.allclose(a.dot(x), b)
+
+
+class TestPermmutterHvp(TfGraphTestCase):
+    """Test class for PermmutterHvp"""
+
+    def test_perm_mutter_hvp_1x1(self):
+        """Test Hessian-vector product for a function with one variable."""
+        policy = HelperPolicy(n_vars=1)
+        x = policy.get_params()[0]
+        a_val = np.array([5.0])
+        a = tf.constant([0.0])
+        f = a * (x**2)
+        expected_hessian = 2 * a_val
+        vector = np.array([10.0])
+        expected_hvp = expected_hessian * vector
+        reg_coeff = 1e-5
+        hvp = PerlmutterHvp()
+
+        self.sess.run(tf.global_variables_initializer())
+        hvp.update_hvp(f, policy, (a, ), reg_coeff)
+        hx = hvp.build_eval(np.array([a_val]))
+        computed_hvp = hx(vector)
+        np.allclose(computed_hvp, expected_hvp)
+
+    @pytest.mark.parametrize('a_val, b_val, x_val, y_val, vector', [
```

Comment:
is there a reason the 1x1 case can't be a case here? shouldn't the mathematical invariant still apply?

Reply:
The test logic depends on the number of function parameters. In 1x1 case, number of function parameters are 1, and for 2x2 its 2. It is easier to have them as separate tests.